### PR TITLE
drivers: add support for MTDs emulated in RAM

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR adds a driver to provide MTDs that are emulated in RAM. It allows to test MTD-based applications on boards that do not provide MTDs by hardware.

It includes also some small documentation fixes for `mtd.h` that were found while writing the driver.

### Testing procedure

The following tests should work on any board:

`tests/pkg_littlefs`
`tests/pkg_littlefs2`
`tests/pkg_spiffs`

### CODE REVIEW

File: drivers/mtd_emulated/Kconfig [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+config MTD_SDCARD_ERASE
+    bool "Enable SD card erase"
+    help
+        Enable this to erase sector before a data write operation.
+        SDCards handle sector erase internally so it's
+        possible to directly write to the card without erasing
+        the sector first hence this feature is disabled by default.
+
+endif # KCONFIG_USEMODULE_MTD_SDCARD
```
Comment (dylad) [inline drivers/mtd_emulated/Kconfig:32]: MTD_SDCARD ?

File: drivers/include/mtd.h [hunk 2]
```diff
@@ -325,7 +325,7 @@ int mtd_read(mtd_dev_t *mtd, void *dest, uint32_t addr, uint32_t count);
  * @param[in]  size     the number of bytes to read
  *
  * @return 0 on success
- * @return < 0 if an error occurred
+ * @return number of bytes read on success
```
Comment (dylad) [inline drivers/include/mtd.h:328]: One of these lines needs to be removed.

File: drivers/include/mtd.h [hunk 3]
```diff
@@ -374,7 +374,7 @@ int mtd_write(mtd_dev_t *mtd, const void *src, uint32_t addr, uint32_t count);
  * @param[in]  size     the number of bytes to write
  *
  * @return 0 on success
- * @return < 0 if an error occurred
+ * @return bytes written on success
```
Comment (dylad) [inline drivers/include/mtd.h:377]: Same here.

File: drivers/include/mtd.h [hunk 4]
```diff
@@ -404,7 +404,7 @@ int mtd_write_page_raw(mtd_dev_t *mtd, const void *src, uint32_t page,
...
```
Comment (dylad) [inline drivers/include/mtd.h:407]: And here.

File: drivers/mtd_emulated/Kconfig [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) 2020 Freie Universitaet Berlin
+#               2020 HAW Hamburg
```
Comment (dylad) [inline drivers/mtd_emulated/Kconfig:2]: Copyright needs update.

**Review by dylad — COMMENTED** [review]
Quick round of review.
I didn't test it yet.

File: drivers/mtd_emulated/Kconfig (hunk 1 above)
Comment (gschorcht) [reply drivers/mtd_emulated/Kconfig:32]: Ups, just a leftover from when I started with a copy. But it is not required at all.

File: drivers/include/mtd.h (hunk 4 above)
Comment (gschorcht) [reply drivers/include/mtd.h:407]: Shit, I removed the wrong line.

File: drivers/mtd_emulated/Kconfig (hunk 5 above)
Comment (gschorcht) [reply drivers/mtd_emulated/Kconfig:2]: I removed the file. It's not included anyway.

**Comment by dylad** [comment]
@gschorcht Please squash !

File: tests/usbus_msc/README.md [hunk 6]
```diff
@@ -39,6 +39,19 @@ by attaching it to the host with the command:
 At this point, the USBUS stack will export the selected MTD devices.
 Devices should appears under /dev/sdX entries.
 
+If the board does not provide a MTD device, a MTD device can be emulated in RAM
+for testing using module `mtd_emulated`:
+```
+USEMODULE=mtd_emulated BAORD=... make ...
```
Comment (dylad) [inline tests/usbus_msc/README.md:45]:
```suggestion
USEMODULE=mtd_emulated BOARD=... make ...
```

File: tests/usbus_msc/README.md [hunk 7]
```diff
@@ -39,6 +39,19 @@ by attaching it to the host with the command:
...
+```
+The emulated MTD device has by default 64 sectors with 4 pages of 128 bytes
+each, the minimum size to be able to create a partition with FAT file system.
+These default parameters can be overridden e.g. by `SECTOR_COUNT`,
+`PAGES_PER_SECTOR` and `PAGE_SIZE`:
+```
+CFLAGS='-DSECTOR_COUNT=128' USEMODULE=mtd_emulated BAORD=... make ...
```
Comment (dylad) [inline tests/usbus_msc/README.md:52]:
```suggestion
CFLAGS='-DSECTOR_COUNT=128' USEMODULE=mtd_emulated BOARD=... make ...
```

**Review by dylad — COMMENTED** [review]
spotted this on lastest commit

File: tests/usbus_msc/README.md (hunk 6 above)
Comment (gschorcht) [reply tests/usbus_msc/README.md]: Good catch. May I squash?

File: drivers/include/mtd_default.h [hunk 8]
```diff
@@ -18,27 +18,52 @@
...
+#define MTD_EMULATED_NUMOF  IS_USED(MODULE_MTD_EMULATED)
+
+/**
+ * @brief   Number of MTD devices
+ */
+#define MTD_NUMOF           (MTD_BOARD_NUMOF + MTD_SDCARD_NUMOF + MTD_EMULATED_NUMOF)
+
+#endif /*  !defined(MTD_NUMOF) && !DOXYGEN */
+
+
```
Comment (dylad) [inline drivers/include/mtd_default.h:60]:
```suggestion
```

File: drivers/include/mtd_emulated.h [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+ * ```
+ * MTD_EMULATED_DEV(0, 16, 4, 64)
+ * ```
+ * creates the emulated MTD device `mtd_emulated_dev0` with 16 sectors, 4 pages
+ * per sector and a page size of 64 bytes. It also defines it as a generic
+ * MTD `mtd2`.
+ *
+ * To use the emulated MTD with the `mtd_default` module, for example, the
+ * `MTD_2` symbol must also be defined before the `mtd_default.h` header file
+ * is included.
```
Comment (dylad) [inline drivers/include/mtd_emulated.h:44]: Why is there a reference to MTD_2 and mtd2 here ? 
This macro should create a `mtd_emulated_t` type

**Review by dylad — COMMENTED** [review]
Please squash (including these two nitpicks)

File: drivers/include/mtd_emulated.h (hunk 9 above)
Comment (gschorcht) [reply drivers/include/mtd_emulated.h]: With the changes in commit https://github.com/RIOT-OS/RIOT/pull/19443/commits/a03d4163a15e845e750e2108e805de9470eccb1b it is no longer necessary to define the emulated MTD also as `MTD_x` macro to get it working with `mtd_default_get_dev`. Therefore, I removed the macro `MTD_EMULATED_DEV_GENERIC` and all the documentation regarding the definition of a generic `mtdx` variable and the `MTD_x` macro.

**Comment by gschorcht** [comment]
Squashed

File: tests/unittests/tests-mtd/tests-mtd.c [hunk 10]
```diff
@@ -43,85 +46,9 @@
...
-    return 0;
-}
-
-static int _power(mtd_dev_t *dev, enum mtd_power_state power)
-{
-    (void)dev;
-    (void)power;
-    return 0;
-}
+MTD_EMULATED_DEV(0, SECTOR_COUNT, PAGE_PER_SECTOR, PAGE_SIZE, WRITE_SIZE);
```
Comment (dylad) [inline tests/unittests/tests-mtd/tests-mtd.c:49]:
```suggestion
MTD_EMULATED_DEV(0, SECTOR_COUNT, PAGE_PER_SECTOR, PAGE_SIZE);
```

So CI will be happy :)

**Review by dylad — CHANGES_REQUESTED** [review]
CI complains about unittests.

File: tests/unittests/tests-mtd/tests-mtd.c (hunk 10 above)
Comment (gschorcht) [reply tests/unittests/tests-mtd/tests-mtd.c:49]: Oh, I forgot to remove an intermediate change that I had introduced especially for the unittest, but then didn't use after all. Fixed.

**Review by dylad — APPROVED** [review]
ACK.

**Comment by dylad** [comment]
bors merge

**Comment by dylad** [comment]
bors merge

**Pull request merged**